core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1552,30 +1560,45 @@ class KafkaApis(val requestChannel: RequestChannel,
         sendResponseMaybeThrottle(request, requestThrottleMs =>
           new AddPartitionsToTxnResponse(requestThrottleMs, partitionErrors.asJava))
       } else {
-        // Send response callback
-        def sendResponseCallback(error: Errors): Unit = {
-          def createResponse(requestThrottleMs: Int): AbstractResponse = {
-            val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(requestThrottleMs,
-              partitionsToAdd.asScala.map{tp => (tp, error)}.toMap.asJava)
-            trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-            responseBody
+        val (_, incorrectMessageFormat) = partitionsToAdd.asScala.partition { tp =>
+          replicaManager.getMagic(tp) match {
```

Comment:
@hachikuji on second thoughts. I think there is one problem with your suggested approach. By the time we send the WriteTxnMarkers request from TC we have already responded to the client with Errors.NONE. So the client would think the commit was successful.